Code/GraphMol/Descriptors/MolData3Ddescriptors.cpp

```diff
@@ -0,0 +1,220 @@
+   int numAtoms= mol.getNumAtoms();
+	double* absionpol=data3D.getIonPOL();
+
+  std::vector<double> ionpols(numAtoms, 0.0);
+  for( int i=0; i<numAtoms; ++i){
+
+    ionpols[i]=absionpol[mol.getAtomWithIdx(i)->getAtomicNum()-1]/absionpol[5];
+  }
+  return ionpols;
+}
+
+
+std::vector<double> MolData3Ddescriptors::GetCharges(const RDKit::ROMol &mol) {
+  std::vector<double> charges(mol.getNumAtoms(), 0);
+  // use 12 iterations... can be more
+  RDKit::computeGasteigerCharges(mol, charges, 12, true);
+  return charges;
+}
+
+int MolData3Ddescriptors::GetPrincipalQuantumNumber(int AtomicNum) {
+  if (AtomicNum<=2) return 1;
+  else if (AtomicNum<=10) return 2;
+  else if (AtomicNum<=18) return 3;
+  else if (AtomicNum<=36) return 4;
+  else if (AtomicNum<=54) return 5;
+  else if (AtomicNum<=86) return 6;
+  else return 7;
+}
+
+
+std::vector<double> MolData3Ddescriptors::GetIState(const  RDKit::ROMol &mol){
+  int numAtoms = mol.getNumAtoms();
+  std::vector<double> Is;
+
+  for (int i = 0; i < numAtoms; ++i) {
+    const RDKit::Atom * atom= mol.getAtomWithIdx(i);
+    int atNum=atom->getAtomicNum();
+    int degree = atom->getDegree();
+    if (degree>0 and atNum>1) {
+      int h = atom->getTotalNumHs();
+      int Zv = RDKit::PeriodicTable::getTable()->getNouterElecs(atNum);
+      double dv =(double) Zv-h;
+      dv = dv / (double) (atNum-Zv-1);
+      int N =  GetPrincipalQuantumNumber(atNum);
+      Is.push_back(round(1000*(4.0/(N*N)*dv+1.0)/degree)/1000);  // WHIM-P5.pdf paper 1997  => +7 & NoHydrogens is used!
+    }
+    else Is.push_back(0);
+ }
+
+ return Is;
+}
+
+
+// adaptation from EState.py 
+// we need the Is value only there
+std::vector<double> MolData3Ddescriptors::GetEState(const  RDKit::ROMol &mol){
+  int numAtoms = mol.getNumAtoms();
+  std::vector<double> Is;
+
+  for (int i = 0; i < numAtoms; ++i) {
+    const RDKit::Atom * atom= mol.getAtomWithIdx(i);
+    int atNum=atom->getAtomicNum();
+    int d = atom->getDegree();
+    if (d>0 and atNum>1) {
+      int h = atom->getTotalNumHs();
+      int dv = RDKit::PeriodicTable::getTable()->getNouterElecs(atNum)-h;
+      int N = GetPrincipalQuantumNumber(atNum); // period value
+      Is.push_back(round(1000*(4.0/(N*N)*dv+1.0)/d)/1000);  // WHIM-P5.pdf paper 1997  => +7 & NoHydrogens is used!
+    }
+    else Is.push_back(1.0); // Is it zero or one for H!
+ }
+
+  double tmp,p;
+  double *dist =  RDKit::MolOps::getDistanceMat(mol,false,false);
+  double accum[numAtoms];
+  for (int i=0;i<numAtoms;i++) {
+    for (int j=i+1;j<numAtoms;j++) {
+       p = dist[i * numAtoms + j]+1;
+       if (p < 1e6) {
+        tmp = (Is[i] - Is[j]) / (p * p);
+        accum[i] += tmp;
+        accum[j] -= tmp;
+      }
+    }
+  }
+
+ for (int i=0;i<numAtoms;i++) {
+    Is[i]+=accum[i];
+ }
+
+ return Is;
+}
+
+// modification of previous code to follow documentation from Padel code
+std::vector<double> MolData3Ddescriptors::GetEState2(const  RDKit::ROMol &mol){
```

Comment:
we need both cause WHIM descriptors need to be centered which is not the case for the others descriptors. Publications are very not easy to decrypt there... and there are some divergences depending of the target descriptor we need to compute. We can add a option and group the code together with a if ... else